overlord/assertstate/assertstate_test.go

```diff
@@ -994,6 +1157,112 @@ func (s *assertMgrSuite) TestRefreshSnapDeclarationsWithStore(c *C) {
 	c.Check(a.(*asserts.Store).Location(), Equals, "the-cloud")
 }
 
+func (s *assertMgrSuite) TestRefreshSnapDeclarationsDownloadError(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	s.setModel(sysdb.GenericClassicModel())
+
+	snapDeclFoo := s.snapDecl(c, "foo", nil)
+
+	s.stateFromDecl(c, snapDeclFoo, "", snap.R(7))
+
+	// previous state
+	err := assertstate.Add(s.state, s.storeSigning.StoreAccountKey(""))
+	c.Assert(err, IsNil)
+	err = assertstate.Add(s.state, s.dev1Acct)
+	c.Assert(err, IsNil)
+	err = assertstate.Add(s.state, snapDeclFoo)
+	c.Assert(err, IsNil)
+
+	// one changed assertion
+	headers := map[string]interface{}{
+		"series":       "16",
+		"snap-id":      "foo-id",
+		"snap-name":    "fo-o",
+		"publisher-id": s.dev1Acct.AccountID(),
+		"timestamp":    time.Now().Format(time.RFC3339),
+		"revision":     "1",
+	}
+	snapDeclFoo1, err := s.storeSigning.Sign(asserts.SnapDeclarationType, headers, nil, "")
+	c.Assert(err, IsNil)
+	err = s.storeSigning.Add(snapDeclFoo1)
+	c.Assert(err, IsNil)
+
+	s.fakeStore.(*fakeStore).downloadAssertionsErr = errors.New("download error")
+
+	err = assertstate.RefreshSnapDeclarations(s.state, 0)
+	c.Assert(err, ErrorMatches, `cannot refresh snap-declarations for snaps:
+ - foo: download error`)
+}
+
+func (s *assertMgrSuite) TestRefreshSnapDeclarationsPersistentNetworkError(c *C) {
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	s.setModel(sysdb.GenericClassicModel())
+
+	snapDeclFoo := s.snapDecl(c, "foo", nil)
+
+	s.stateFromDecl(c, snapDeclFoo, "", snap.R(7))
+
+	// previous state
+	err := assertstate.Add(s.state, s.storeSigning.StoreAccountKey(""))
+	c.Assert(err, IsNil)
+	err = assertstate.Add(s.state, s.dev1Acct)
+	c.Assert(err, IsNil)
+	err = assertstate.Add(s.state, snapDeclFoo)
+	c.Assert(err, IsNil)
+
+	// one changed assertion
+	headers := map[string]interface{}{
+		"series":       "16",
+		"snap-id":      "foo-id",
+		"snap-name":    "fo-o",
+		"publisher-id": s.dev1Acct.AccountID(),
+		"timestamp":    time.Now().Format(time.RFC3339),
+		"revision":     "1",
+	}
+	snapDeclFoo1, err := s.storeSigning.Sign(asserts.SnapDeclarationType, headers, nil, "")
+	c.Assert(err, IsNil)
+	err = s.storeSigning.Add(snapDeclFoo1)
+	c.Assert(err, IsNil)
+
+	pne := new(httputil.PersistentNetworkError)
+	s.fakeStore.(*fakeStore).snapActionErr = pne
+
+	err = assertstate.RefreshSnapDeclarations(s.state, 0)
+	c.Assert(err, Equals, pne)
+}
+
+func (s *assertMgrSuite) TestRefreshSnapDeclarationsNoStoreFallback(c *C) {
+	// test that if we get a 4xx or 500 error from the store trying bulk
+	// assertion refresh we fall back to the old logic
+	s.fakeStore.(*fakeStore).snapActionErr = &store.UnexpectedHTTPStatusError{StatusCode: 400}
+
+	s.TestRefreshSnapDeclarationsNoStore(c)
+}
```

Comment:
Could we check that the error condition was actually hit or is it too annoying? At the very least we could inspect logger for "bulk refresh of snap-declarations failed.." message?